src/libsyntax/parse/parser.rs

```diff
@@ -580,12 +580,21 @@ impl<'a> Parser<'a> {
                 self.bug("ident interpolation not converted to real token");
             }
             _ => {
-                let mut err = self.fatal(&format!("expected identifier, found `{}`",
-                                                  self.this_token_to_string()));
-                if self.token == token::Underscore {
-                    err.note("`_` is a wildcard pattern, not an identifier");
-                }
-                Err(err)
+                let last_token = self.last_token.clone().map(|t| *t);
+                Err(match last_token {
```

Comment:
At this point `last_token` is the actual doc comment, but the failure is raised by the current `self.token` not being a valid element at this location, that's why we have to look back to check wether the failure is because the previous element was a doc comment or something else. This is also why we set error to `self.last_span` instead of `self.span`.

For example, with the following input `file.rs`:
```rust
fn foo(a: u8, b: u64) {
    /// document
}
```
The type of `self.last_token` is `Some(DocComment(/// document(61)))` while `self.token` is `CloseDelim(Brace)` and you get the following output with the current code:

```bash
% ./rustc file.rs
file.rs:2:5: 2:17 error: found a documentation comment that doesn't document anything
file.rs:2     /// document
              ^~~~~~~~~~~~
file.rs:2:5: 2:17 help: doc comments must come before what they document, maybe a comment was intended with `//`?
```